src/WhereHasConditions/WhereHasConditionsDirective.php

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+
+        return $builder;
+    }
+
+    public static function invalidColumnName(string $column): string
+    {
+        return "Column names may contain only alphanumerics or underscores, and may not begin with a digit, got: $column";
+    }
+
+    /**
+     * Manipulate the AST.
+     *
+     * @param  \Nuwave\Lighthouse\Schema\AST\DocumentAST  $documentAST
+     * @param  \GraphQL\Language\AST\InputValueDefinitionNode  $argDefinition
+     * @param  \GraphQL\Language\AST\FieldDefinitionNode  $parentField
+     * @param  \GraphQL\Language\AST\ObjectTypeDefinitionNode  $parentType
+     * @return void
+     */
+    public function manipulateArgDefinition(
+        DocumentAST &$documentAST,
+        InputValueDefinitionNode &$argDefinition,
+        FieldDefinitionNode &$parentField,
+        ObjectTypeDefinitionNode &$parentType
+    ): void {
+        if ($allowedColumns = $this->directiveArgValue('columns')) {
+            $restrictedWhereHasConditionsName = $this->restrictedWhereHasConditionsName($argDefinition, $parentField);
+            $argDefinition->type = PartialParser::namedType($restrictedWhereHasConditionsName);
+
+            $allowedColumnsEnumName = Codegen::allowedColumnsEnumName($argDefinition, $parentField);
+
+            $documentAST
+                ->setTypeDefinition(
+                    WhereHasConditionsServiceProvider::createWhereHasConditionsInputType(
+                        $restrictedWhereHasConditionsName,
+                        "Dynamic relationship WHERE conditions for the `{$argDefinition->name->value}` argument on the query `{$parentField->name->value}`.",
+                        $allowedColumnsEnumName
+                    )
+                )
+                ->setTypeDefinition(
+                    Codegen::createAllowedColumnsEnum($argDefinition, $parentField, $allowedColumns, $allowedColumnsEnumName)
+                );
+        } else {
+            $argDefinition->type = PartialParser::namedType(WhereHasConditionsServiceProvider::DEFAULT_WHERE_HAS_CONDITIONS);
+        }
+    }
+
+    /**
+     * Create the name for the restricted WhereHasConditions input.
+     *
+     * @example FieldNameArgNameWhereHasConditions
+     *
+     * @param  \GraphQL\Language\AST\InputValueDefinitionNode  $argDefinition
+     * @param  \GraphQL\Language\AST\FieldDefinitionNode  $parentField
+     * @return string
+     */
+    protected function restrictedWhereHasConditionsName(InputValueDefinitionNode &$argDefinition, FieldDefinitionNode &$parentField): string
+    {
+        return Str::studly($parentField->name->value)
+            .Str::studly($argDefinition->name->value)
+            .'WhereHasConditions';
+    }
+
+    /**
+     * Ensure the column name is well formed and prevent SQL injection.
+     *
+     * @param  string  $column
+     * @return void
+     *
+     * @throws \GraphQL\Error\Error
+     */
+    protected static function assertValidColumnName(string $column): void
```

Comment:
The `assertValidColumnName` method could be moved to the `SQLOperator` class, so that we don't have to duplicate it in multiple directives.